### PR TITLE
Add SelfServeRrpBeaconServerWhitelister Contract

### PR DESCRIPTION
**Pull request by Ashar2shahid**
Main purpose of this contract is for readers to whitelist themselves. The contract is assumed to have the `indefiniteWhitelisterRole` and `WhitelistExpirationSetterRole`. 

There are two ways to whitelist:
1. With an expirationTimestamp
2. Indefinitely

A set of beaconIds are available for readers to be able to whitelist themselves for either 1 or 2 (different sets). The deployer(owner) can add/remove from the sets. 
 
I wasn't sure where to put the contracts, so I put it under `/utils`.

### CODE REVIEW

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (Siegrift) [inline packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:115]: This needs to be shortened, right?

File: packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js
Comment (Siegrift) [inline packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js:174]: Figuring out this step would take me a lot of time. It was good, that you took over this for me.

File: packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js
Comment (Siegrift) [inline packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js:319]: Shouldn't we have a similar test that tries to whitelist a reader, but doesn't have the permissions for it?

File: packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js
Comment (Siegrift) [inline packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js:55]: Nit: Shouldn't this be moved to some utils file?

File: packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js
Comment (Siegrift) [inline packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js:74]: Nit: Shouldn't this be moved to some utils file?

**Review by Siegrift — APPROVED** [review]
LGTM, but my contract experience is limited so better wait for someone elses approval.

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (acenolaza) [inline packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:111]: So we can't really set the indefinite whitelist status back to `false` using this contract, correct? Wouldn't it be better to have a setter redirect instead?

**Review by acenolaza — COMMENTED** [review]
Do we really need to keep track of the `expirationTimestamps` and  the `indefiniteWhitelistStatuses` in this contract? Shouldn't we only store the `beaconIds` that are allowed to be managed by this contract?

Then when we call the whitelisting function we can pass the `expirationTimestamp` or the `indefiniteWhitelistStatus`.

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (Ashar2shahid) [reply packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:29]: yeah I forgot we don't do this anymore, this was how we used to do it for function argument variables.

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (Ashar2shahid) [reply packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:34]: its usually standard practice to check for address(0) for contract addresses. It comes up in audits so I added the check here

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (Ashar2shahid) [reply packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:115]: I don't follow ?

File: packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js
Comment (Ashar2shahid) [reply packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js:319]: do you mean the contract itself doesn't have the permissions ?

File: packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js
Comment (Ashar2shahid) [reply packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js:55]: I think this is fine to remain here

File: packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js
Comment (Ashar2shahid) [reply packages/airnode-protocol/test/utils/SelfServeRrpBeaconServerWhitelister.sol.js:74]: this one definently can be moved to utils

**Comment by Ashar2shahid** [comment]
> Do we really need to keep track of the `expirationTimestamps` and the `indefiniteWhitelistStatuses` in this contract? Shouldn't we only store the `beaconIds` that are allowed to be managed by this contract?
> 
> Then when we call the whitelisting function we can pass the `expirationTimestamp` or the `indefiniteWhitelistStatus`.

So the purpose of this contract is for readers to be able to whitelist themselves, for the hackathon we want users to be able to whitelist themselves upto an expiration, so we will add those beaconIds with their respective expirationIds to the mapping and allow users to whitelist themselves. 

We also want certain beaconIds to always be available for testing purposes (think testnets) in this case we can add these beaconIds to the indefinite mapping and users of this beaconIds can be whitelisted indefinetely.

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (acenolaza) [reply packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:115]: "Error message for require is too long" lint warning
Maybe just change it to "Cannot whitelist indefinitetly"

File: packages/airnode-protocol/contracts/utils/interfaces/ISelfServeRrpBeaconServerWhitelister.sol
Comment (acenolaza) [inline packages/airnode-protocol/contracts/utils/interfaces/ISelfServeRrpBeaconServerWhitelister.sol:37]: Should we also add the 2 public getters for `beaconIdToExpirationTimestamp` and `beaconIdToIndefiniteWhitelistStatus` mappings?

**Review by acenolaza — APPROVED** [review]

File: packages/airnode-protocol/contracts/utils/interfaces/ISelfServeRrpBeaconServerWhitelister.sol
Comment (Ashar2shahid) [reply packages/airnode-protocol/contracts/utils/interfaces/ISelfServeRrpBeaconServerWhitelister.sol:37]: don't public mappings generate getters automatically?

Comment (acenolaza) [reply packages/airnode-protocol/contracts/utils/interfaces/ISelfServeRrpBeaconServerWhitelister.sol:37]: they do but if a contract only wants to import the interface to call any of the getters then they wouldn't be available

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (bbenligiray) [inline packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:32]: This would always be all `true`s so it's weird. Do this instead:
```
        address rrpBeaconServerAddress,
        bytes32[] memory beaconIds,
        uint64[] memory expirationTimestamps,
        bool[] memory indefiniteWhitelistStatuses
```

Comment (bbenligiray) [reply packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:32]: Also validate that the argument lengths are equal

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (bbenligiray) [inline packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:14]: The point of initializing the parameters in the constructor was that this contract was going to be immutable / we wouldn't need to worry about who the owner is and if the wallet would get compromised. So this contract should either be ownable and not do anything in the constructor, or not be ownable and do everything in the constructor and never again.

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (bbenligiray) [inline packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:91]: I think it's simpler to have a single `whitelistReader()` function that both sets the whitelist expiration and the indefinite whitelist status

**Comment by bbenligiray** [comment]
I don't want this contract in the airnode-protocol package tbh. We will be using a lot of such one-use contracts and we can't clutter the monorepo with them.

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (Ashar2shahid) [reply packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:32]: I see what you mean, more like this then:
```    
bytes32[] memory beaconIdsExpirationTimestamp,
uint64[] memory expirationTimestamps,
bytes32[] memory beaconIdsIndefiniteWhitelistStatus
```

and not bool

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (Ashar2shahid) [reply packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:14]: The owner of this contract is going to be the OwnableCallForwarder after deployment, who is also the manager of the `RequesterAuthorizerWithManager`. So I think its fine for this to be a ownable contract because if that is compromised then the `RequesterAuthorizer` is compromised which is a much bigger issue.

Yes, I agree probably better to remove from the constructor then

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (Ashar2shahid) [reply packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:91]: okay makes sense

File: packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol
Comment (bbenligiray) [reply packages/airnode-protocol/contracts/utils/SelfServeRrpBeaconServerWhitelister.sol:32]: I mean you don't even need two lists of beacons, it could go like
```
[beaconId1, beaconId2, beaconId3]
[12345678, 0, 987654312]
[false, true, false]
```
Also `beaconIdsExpirationTimestamp` and `beaconIdsIndefiniteWhitelistStatus` are bad variable names. It should be something like `temporarilyWhitelistableBeaconIds` / `indefinitelyWhitelistableBeaconIds` (if you're going to go with https://github.com/api3dao/airnode/pull/789#discussion_r792637995)

**Comment by bbenligiray** [comment]
Moved to https://github.com/api3dao/utility-contracts

**Pull request closed without merging**
